File: samples/src/main/java/com/skedgo/android/samples/TripDetailsActivity.java
package com.skedgo.android.samples;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.BottomSheetBehavior;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.skedgo.android.common.model.Location;
import com.skedgo.android.common.model.Query;
import com.skedgo.android.common.model.Trip;
import com.skedgo.android.common.model.TripGroup;
import com.skedgo.android.common.util.TripSegmentUtils;
import com.skedgo.android.tripkit.RouteService;
import com.skedgo.android.tripkit.TripKit;

import java.util.List;

import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.functions.Func2;

public class TripDetailsActivity extends AppCompatActivity {
  private static final String TRIP_EXTRA = "TripExtra";

  public static Intent newIntent(Context context, Trip trip) {
    Intent intent = new Intent(context, TripDetailsActivity.class);
    intent.putExtra(TRIP_EXTRA, trip);
    return intent;
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_trip_details);
    final RecyclerView segmentList = (RecyclerView) findViewById(R.id.segmentList);

    // final Trip trip = getIntent().getParcelableExtra(TRIP_EXTRA);
    Query query = new Query();
    query.setFromLocation(new Location(10.810491, 106.705726));
    query.setToLocation(new Location(10.754426, 106.624733));
    TripKit.singleton().getRouteService()
        .routeAsync(query)
        .flatMap(new Func1<List<TripGroup>, Observable<TripGroup>>() {
          @Override public Observable<TripGroup> call(List<TripGroup> tripGroups) {
            return Observable.from(tripGroups);
          }
        })
        .map(new Func1<TripGroup, Trip>() {
          @Override public Trip call(TripGroup tripGroup) {
            return tripGroup.getDisplayTrip();
          }
        })
        .toSortedList(new Func2<Trip, Trip, Integer>() {
          @Override public Integer call(Trip trip, Trip trip2) {
            return trip.getSegments().size() - trip2.getSegments().size();
          }
        })
        .map(new Func1<List<Trip>, Trip>() {
          @Override public Trip call(List<Trip> trips) {
            return trips.get(trips.size() - 1);
          }
        })
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Action1<Trip>() {
          @Override public void call(final Trip trip) {
            segmentList.setAdapter(new RecyclerView.Adapter<SegmentViewHolder>() {

              @Override public SegmentViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
                return new SegmentViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.segment, parent, false));
              }

              @Override public void onBindViewHolder(SegmentViewHolder holder, int position) {
                holder.actionTitle.setText(TripSegmentUtils.getTripSegmentAction(holder.itemView.getContext(), trip.getSegments().get(position)));
              }

              @Override public int getItemCount() {
                return trip.getSegments().size();
              }
            });
          }
        });
    segmentList.setLayoutManager(new LinearLayoutManager(this));
    BottomSheetBehavior<View> from = BottomSheetBehavior.from(findViewById(R.id.bottomSheet));
    from.setPeekHeight(360);
    // from.setState();
  }

  private static class SegmentViewHolder extends RecyclerView.ViewHolder {
    final TextView actionTitle;

    SegmentViewHolder(View itemView) {
      super(itemView);
      actionTitle = (TextView) itemView.findViewById(R.id.segmentTitle);
    }
  }
}

File: A2bTrips/src/main/java/skedgo/tripkit/demo/a2btrips/TripDetailsActivity.kt
package skedgo.tripkit.demo.a2btrips

import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.BottomSheetBehavior
import android.support.v7.app.AppCompatActivity
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.MarkerOptions
import com.skedgo.android.common.model.Trip
import com.skedgo.android.tripkit.TripLinesProcessor
import rx.Observable
import skedgo.tripkit.demo.a2btrips.databinding.ActivityTripDetailsBinding

class TripDetailsActivity : AppCompatActivity() {

  companion object {
    private val TRIP_EXTRA = "TripExtra"

    fun newIntent(context: Context, trip: Trip): Intent {
      val intent = Intent(context, TripDetailsActivity::class.java)
      intent.putExtra(TRIP_EXTRA, trip)
      return intent
    }
  }

  val trip: Trip by lazy { intent.getParcelableExtra<Trip>(TRIP_EXTRA) }

  val viewModel by lazy { TripDetailsViewModel(this, trip) }
  val tripLinesProcessor = TripLinesProcessor()

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    val binding = DataBindingUtil.setContentView<ActivityTripDetailsBinding>(this, R.layout.activity_trip_details)
    binding.viewModel = viewModel
    val mapFragment = supportFragmentManager.findFragmentById(R.id.mapFragment) as SupportMapFragment
    mapFragment.getMapAsync { map ->
      tripLinesProcessor.getPolylineOptionsListAsync(trip.segments)
          .flatMap { Observable.from(it) }
          .subscribe { map.addPolyline(it) }
      val height = resources.displayMetrics.heightPixels;
      val width = resources.displayMetrics.widthPixels;

      map.animateCamera(CameraUpdateFactory.newLatLngBounds(LatLngBounds.Builder()
          .include(LatLng(trip.from.lat, trip.from.lon))
          .include(LatLng(trip.to.lat, trip.to.lon))
          .build(), width, height, height / 10))

      map.addMarker(MarkerOptions()
          .position(LatLng(trip.from.lat, trip.from.lon))
          .icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_GREEN)))
      map.addMarker(MarkerOptions()
          .position(LatLng(trip.to.lat, trip.to.lon))
          .icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_RED)))
    }
    val bottomSheet = BottomSheetBehavior.from(binding.bottomSheet)
    bottomSheet.peekHeight = 360
  }
}